hugr-passes/src/lower_types/linearize.rs

```diff
@@ -206,19 +219,172 @@ impl Linearizer {
     }
 }
 
+fn runtime_reqs(h: &Hugr) -> ExtensionSet {
+    h.signature(h.root()).unwrap().runtime_reqs.clone()
+}
+
+pub fn discard_array(args: &[TypeArg], lin: &Linearizer) -> Result<OpReplacement, LinearizeError> {
+    // Require known length i.e. usable only after monomorphization, due to no-variables limitation
+    // restriction on OpReplacement::CompoundOp
+    let [TypeArg::BoundedNat { n }, TypeArg::Type { ty }] = args else {
+        panic!("Illegal TypeArgs to array: {:?}", args)
+    };
+    // Make a function that maps the linear element to unit
+    let map_fn = {
+        let mut dfb = DFGBuilder::new(inout_sig(ty.clone(), Type::UNIT)).unwrap();
+        let [to_discard] = dfb.input_wires_arr();
+        lin.discard_op(ty)?.add(&mut dfb, [to_discard]).unwrap();
+        let ret = dfb.add_load_value(Value::unary_unit_sum());
+        dfb.finish_hugr_with_outputs([ret]).unwrap()
+    };
+    // Now array.scan that over the input array to get an array of unit (which can be discarded)
+    let array_scan = ArrayScan::new(ty.clone(), Type::UNIT, vec![], *n, runtime_reqs(&map_fn));
+    let in_type = array_type(*n, ty.clone());
+    Ok(OpReplacement::CompoundOp(Box::new({
+        let mut dfb = DFGBuilder::new(inout_sig(in_type, type_row![])).unwrap();
+        let [in_array] = dfb.input_wires_arr();
+        let map_fn = dfb.add_load_value(Value::Function {
+            hugr: Box::new(map_fn),
+        });
+        // scan has one output, an array of unit, so just ignore/discard that
+        dfb.add_dataflow_op(array_scan, [in_array, map_fn]).unwrap();
+        dfb.finish_hugr_with_outputs([]).unwrap()
+    })))
+}
+
+pub fn copy_array(args: &[TypeArg], lin: &Linearizer) -> Result<OpReplacement, LinearizeError> {
+    // Require known length i.e. usable only after monomorphization, due to no-variables limitation
+    // restriction on OpReplacement::CompoundOp
+    let [TypeArg::BoundedNat { n }, TypeArg::Type { ty }] = args else {
+        panic!("Illegal TypeArgs to array: {:?}", args)
+    };
+    let i64_t = INT_TYPES[6].to_owned();
+    let option_sty = option_type(ty.clone());
+    let option_ty = Type::from(option_sty.clone());
+    let option_array = array_type(*n, option_ty.clone());
+    let copy_elem = {
```

Comment:
Maybe add a comment on top explaining what this function does, or how we go about copying in general?